specs/heading_attrs.txt

```diff
@@ -0,0 +1,554 @@
+````````````````````````````````
+
+Right curly brace:
+
+```````````````````````````````` example
+# H1 {.foo}bar}
+.
+<h1>H1 {.foo}bar}</h1>
+````````````````````````````````
+
+less-than operator and greater-than operator:
+
+```````````````````````````````` example
+# H1 {<i>foo</i>}
+.
+<h1>H1 {<i>foo</i>}</h1>
+````````````````````````````````
+
+Backslash:
+
+```````````````````````````````` example
+# H1 {.foo\}
+.
+<h1>H1 {.foo}</h1>
+````````````````````````````````
+
+Newline:
+
+```````````````````````````````` example
+H1 {.foo
+.bar}
+==
+.
+<h1>H1 {.foo
+.bar}</h1>
+````````````````````````````````
+
+# Cancelling parsing of attribute blocks
+
+If you want to avoid the last `{...}` from being treated as an attribute block,
+you can do it in some ways.
+
+Put a dummy empty attribute block `{}` at the end of line:
+
+```````````````````````````````` example
+H1 {} {}
+=====
+
+## H2 {} {}
+.
+<h1>H1 {}</h1>
+<h2>H2 {}</h2>
+````````````````````````````````
+
+Or, put closing hashes for ATX heading:
+
+```````````````````````````````` example
+## H2 {} ##
+.
+<h2>H2 {}</h2>
+````````````````````````````````
+
+Or, include forbidden characters (a backslash and newline in this example)
+between the braces.
+
+```````````````````````````````` example
+# H1 {\}
+## this is also ok \{\}
+
+newline can be used for setext heading {
+}
+--
+.
+<h1>H1 {}</h1>
+<h2>this is also ok {}</h2>
+<h2>newline can be used for setext heading {
+}</h2>
+````````````````````````````````
+
+Note that escaping only opening braces does not prevent the braces from
+starting attribute blocks.
+Forbidden characters between braces are important.
+
+```````````````````````````````` example
+# H1 \{.foo}
+## H2 \\{.bar}
+.
+<h1 class="foo">H1 </h1>
+<h2 class="bar">H2 \</h1>
+````````````````````````````````
+<!--
+  Note that the `\` at `H1-2 \</h1>` should be a single backslash, not double.
+  This is due to how `<Spec as Iterator>::next()` in build.rs parses this file.
+  The file is read as a plain string and the tests cases are generated.
+  The lines after a period in examples are not parsed as a markdown.
+-->
+
+# Disabled inlines
+
+Inlines are disabled and special characters such as `_` and `*` will be treated
+as a plain character inside attribute blocks.
+
+```````````````````````````````` example
+# H1 {#`code`}
+## H2 {#foo__bar__baz}
+### H3 {#foo**bar**baz}
+.
+<h1 id="`code`">H1</h1>
+<h2 id="foo__bar__baz">H2</h2>
+<h3 id="foo**bar**baz">H3</h3>
+````````````````````````````````
+
+```````````````````````````````` example
+H1 {#`code`}
+==
+
+H2-1 {#foo__bar__baz}
+----
+
+H2-2 {#foo**bar**baz}
+--
+.
+<h1 id="`code`">H1</h1>
+<h2 id="foo__bar__baz">H2-1</h2>
+<h2 id="foo**bar**baz">H2-2</h2>
+````````````````````````````````
+
+Parsing of attribute blocks takes precedence, even when unclosed inlines are
+present before the attribute blocks.
+
+````````````````````````````````.example
+# H1__ {#my__id1}
+## H2** {#my**id2}
+### H3 ` {.code` }
```

Comment:
This is an excellent test case. I do wonder if we are in line with other parsers though?

Reply:
[babelmark3 | Compare Markdown Implementations](https://babelmark.github.io/?text=%23+foo__+%7B%23my__id1%7D%0A%23+bar+__baz+%7B%23my__id2%7D%0A%0A__qux)

* Same:
    + markdig 0.26.0.0 and php-markdown-extra 1.9.0 converts to the same result as this PR.
* Similar:
    + maruku 0.7.3.beta1 seems to try to do the same, but fails with handling of the unclosed inlines (unclosed `__baz` and `__qux` are converted to `<strong>...`).
* Different:
    + cebe/extra 1.2.0, pandoc 2.16.2, and kramdown 1.2.0 gave up on `# bar __baz {#my__id2}` case.
    + cebe/extra converts `__baz {#my__` into `<strong>...</strong>`.
    + pandoc seems to fail to convert inlines, so `__baz {#my__` is written as is.
    + kramdown even fails to make it a heading.

There seems no "common" way to handle this, but there are roughly two patterns: "blocks are trimmed first" and "give up when ambiguous".
(I'm a bit surprised that there is no "inlines takes precedence" pattern.)

I think it is better to use simple rule "attribute blocks are trimmed first", since:

* Class names such as `.foo__bar` would be widely used and I want them to be writable as `{ .foo__bar }`.
    + I don't think it is good if programatically added `{.foo__bar}` makes the parsing "ambiguous".
* "Give up" patterns have variations, and it is less clear what (should) happens when "ambiguous".